File: projects/CPP/classes/SoccerDecoder.cpp
#include "../headers/SoccerDecoder.hpp"

#define SQL_LAYOUTA "SELECT `layout` FROM `teams` WHERE `id` LIKE 1"
#define SQL_LAYOUTB "SELECT `layout` FROM `teams` WHERE `id` LIKE 2"
#define SQL_PLAYERSA "SELECT * FROM `players` WHERE `team_id` LIKE 1 ORDER BY `rating` ASC LIMIT 11"
#define SQL_PLAYERSB "SELECT * FROM `players` WHERE `team_id` LIKE 2 ORDER BY `rating` ASC LIMIT 11"
#define SQL_USE_DB "USE soccer"

SoccerDecoder::SoccerDecoder(int argc, char* argv[]) {
	auto t = std::chrono::high_resolution_clock::now();
	unsigned int seed = std::chrono::system_clock::now().time_since_epoch().count();
	this->threadCount = std::thread::hardware_concurrency();
	this->duration = std::chrono::duration_cast<std::chrono::seconds>(t - t);
	if (argc >= 3) {
		DatabaseType x = static_cast<enum class DatabaseType>(std::stoi(argv[2]));
		this->faceoffs = std::stoi(argv[1]);
		switch (x) { // NOTE: Original idea is every player will have pre determined position but I wanted to get this part quickly done, I might change it later
			case DatabaseType::SQLITE : { 
				DB::LiteDriver db(argv[3]);

				db.execSingleStmt(SQL_LAYOUTA, true);
				int layoutA = db.getIntegerFromRow(db.getStorageRefference(), "layout", 0);
				this->layoutA = static_cast<Match::Team::Formations>(layoutA);

				db.execSingleStmt(SQL_LAYOUTB, true);
				int layoutB = db.getIntegerFromRow(db.getStorageRefference(), "layout", 0);
				this->layoutB = static_cast<Match::Team::Formations>(layoutB);

				db.execSingleStmt(SQL_PLAYERSA, true);
				this->SetPlayersA(this->FillPlayersSQLITE(db, layoutA));
				this->SetGoalKeeperA(FieldUnit::Player(db.getDoubleFromRow(db.getStorageRefference(), "rating", 10), db.getDoubleFromRow(db.getStorageRefference(), "stamina", 10), db.getDoubleFromRow(db.getStorageRefference(), "age", 10), db.getDoubleFromRow(db.getStorageRefference(), "experience", 10), db.getStringFromRow(db.getStorageRefference(), "name", 10), FieldUnit::Player::PositionType::GOALKEEPER));

				db.execSingleStmt(SQL_PLAYERSB, true);
				this->SetPlayersB(this->FillPlayersSQLITE(db, layoutB));
				this->SetGoalKeeperB(FieldUnit::Player(db.getDoubleFromRow(db.getStorageRefference(), "rating", 10), db.getDoubleFromRow(db.getStorageRefference(), "stamina", 10), db.getDoubleFromRow(db.getStorageRefference(), "age", 10), db.getDoubleFromRow(db.getStorageRefference(), "experience", 10), db.getStringFromRow(db.getStorageRefference(), "name", 10), FieldUnit::Player::PositionType::GOALKEEPER));
				break;
			}
			case DatabaseType::MARIADB:
			case DatabaseType::MYSQL: {
				sql::Connection* conn = nullptr;
				sql::mysql::MySQL_Driver* driver = nullptr;
				sql::Statement* stmt = nullptr;
				sql::ResultSet* resLayout = nullptr;
				sql::ResultSet* resPlayers = nullptr;
				try{
					driver = sql::mysql::get_mysql_driver_instance();
					if (x == DatabaseType::MARIADB) { // Could be in config/argv, will probably move it later on when I have time to cleanup
						conn = driver->connect("tcp://127.0.0.1:3306", "root", "securepassword");
					}
					else {
						conn = driver->connect("tcp://127.0.0.1:4306", "root", "securepassword");
					}
					stmt = conn->createStatement();
					stmt->execute(SQL_USE_DB);
					resPlayers = stmt->executeQuery(SQL_PLAYERSA);
					resLayout = stmt->executeQuery(SQL_LAYOUTA);
					resLayout->first();
					this->SetPlayersA(this->FillPlayersMYSQL(resPlayers, resLayout->getInt("layout")));
					resPlayers->last();
					this->SetGoalKeeperA(FieldUnit::Player(static_cast<double>(resPlayers->getDouble("rating")), static_cast<double>(resPlayers->getDouble("stamina")), static_cast<double>(resPlayers->getDouble("age")), static_cast<double>(resPlayers->getDouble("experience")), resPlayers->getString("name"), FieldUnit::Player::PositionType::GOALKEEPER));

					resPlayers = stmt->executeQuery(SQL_PLAYERSB);
					resLayout = stmt->executeQuery(SQL_LAYOUTB);
					resLayout->first();
					this->SetPlayersB(this->FillPlayersMYSQL(resPlayers, resLayout->getInt("layout")));
					resPlayers->last();
					this->SetGoalKeeperB(FieldUnit::Player(static_cast<double>(resPlayers->getDouble("rating")), static_cast<double>(resPlayers->getDouble("stamina")), static_cast<double>(resPlayers->getDouble("age")), static_cast<double>(resPlayers->getDouble("experience")), resPlayers->getString("name"), FieldUnit::Player::PositionType::GOALKEEPER));
				}
				catch (std::exception& e){
					std::cout << e.what();
					exit(-1);
				}
				break;
			}
		}
	} else { // Default sample data
		this->layoutA = Match::Team::Formations::_4_3_3;
		this->layoutB = Match::Team::Formations::_4_3_3;

		this->SetPlayersA({
			FieldUnit::Player(8.5, 10.0, 34, 1.0, "Kanker", FieldUnit::Player::PositionType::MIDFIELDER),
			FieldUnit::Player(6.5, 10.0, 34, 1.0, "asda", FieldUnit::Player::PositionType::MIDFIELDER),
			FieldUnit::Player(6.5, 10.0, 34, 10.0, "Epfdgfdgfgd", FieldUnit::Player::PositionType::MIDFIELDER),
			FieldUnit::Player(7.5, 10.0, 34, 40.0, "Epxcxxcxcic", FieldUnit::Player::PositionType::MIDFIELDER),
			FieldUnit::Player(8.5, 10.0, 34, 4.0, "Eqwqwwqwqwpic", FieldUnit::Player::PositionType::STRIKERS),
			FieldUnit::Player(8.5, 10.0, 34, 40.0, "Epicgfdgfdgfdg", FieldUnit::Player::PositionType::STRIKERS),
			FieldUnit::Player(6.5, 10.0, 34, 40.0, "Eczxxczxczxczpic", FieldUnit::Player::PositionType::STRIKERS),
			FieldUnit::Player(0.5, 10.0, 34, 4.0, "Epqqqqqqqqqic", FieldUnit::Player::PositionType::DEFENDER),
			FieldUnit::Player(4.5, 10.0, 34, 4.0, "Epaaaaaaaaic", FieldUnit::Player::PositionType::DEFENDER),
			FieldUnit::Player(8.5, 10.0, 34, 40.0, "Epzzzzzzzzzzic", FieldUnit::Player::PositionType::DEFENDER) });
		this->SetGoalKeeperA(FieldUnit::Player(9.5, 10.0, 34, 74.0, "Pot", FieldUnit::Player::PositionType::GOALKEEPER));

		this->SetPlayersB({
			FieldUnit::Player(8.5, 10.0, 34, 1.0, "Kanker", FieldUnit::Player::PositionType::MIDFIELDER),
			FieldUnit::Player(8.5, 10.0, 34, 1.0, "asda", FieldUnit::Player::PositionType::MIDFIELDER),
			FieldUnit::Player(6.5, 10.0, 34, 10.0, "Epfdgfdgfgd", FieldUnit::Player::PositionType::MIDFIELDER),
			FieldUnit::Player(8.5, 6.0, 34, 4.0, "Epxcxxcxcic", FieldUnit::Player::PositionType::MIDFIELDER),
			FieldUnit::Player(8.5, 10.0, 34, 4.0, "Eqwqwwqwqwpic", FieldUnit::Player::PositionType::STRIKERS),
			FieldUnit::Player(7.5, 10.0, 34, 4.0, "Epicgfdgfdgfdg", FieldUnit::Player::PositionType::STRIKERS),
			FieldUnit::Player(8.5, 10.0, 34, 4.0, "Eczxxczxczxczpic", FieldUnit::Player::PositionType::STRIKERS),
			FieldUnit::Player(8.5, 7.0, 34, 4.0, "Epqqqqqqqqqic", FieldUnit::Player::PositionType::DEFENDER),
			FieldUnit::Player(5.5, 10.0, 34, 4.0, "Epaaaaaaaaic", FieldUnit::Player::PositionType::DEFENDER),
			FieldUnit::Player(8.5, 10.0, 34, 4.0, "Epzzzzzzzzzzic", FieldUnit::Player::PositionType::DEFENDER) });
		this->SetGoalKeeperB(FieldUnit::Player(7.5, 10.0, 34, 74.0, "Nod", FieldUnit::Player::PositionType::GOALKEEPER));

		this->faceoffs = DEFAULT_MATCHES;
	}
}

void SoccerDecoder::Start() {
	auto start = std::chrono::high_resolution_clock::now();
	std::vector<std::thread> threads;

	if (!(this->faceoffs < this->threadCount * 2)) {
		for (auto& var : Chunks(this->faceoffs))
			threads.push_back(std::thread([=]
				{ Func(var); }));
		for (auto& t : threads)
			if (t.joinable())
				t.join();
	}
	else Func(this->faceoffs);

	this->SetDuration(std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - start));
}

void SoccerDecoder::Func(int chunk)
{
	srand(time(NULL));

	Match::Team a("Attackers", Match::Team::FieldType::AWAY, this->GetPlayersA(), this->GetGoalKeeperA(), this->layoutA);
	Match::Team b("Defenders", Match::Team::FieldType::HOME, this->GetPlayersB(), this->GetGoalKeeperB(), this->layoutB);

	for (size_t i = 0; i < chunk; i++) {
		Match::FaceOff PlayOff(a, b);

		PlayOff.Simulate();

		if (PlayOff.GetAttackingTeam().GetScore() > PlayOff.GetDefendingTeam().GetScore()) aw++;
		else if (PlayOff.GetAttackingTeam().GetScore() < PlayOff.GetDefendingTeam().GetScore()) bw++;
		else dw++;
	}
}

std::vector<FieldUnit::Player> SoccerDecoder::FillPlayersSQLITE(DB::LiteDriver& db, int layout)
{
	int row = 0;
	std::vector<FieldUnit::Player> players;
	for (int i = 0; i < 3; ++i) {
		int temp = layout % 10;
		for (int j = 0; j < temp; ++j) {
			switch (i) {
			case 0: {
				players.push_back(FieldUnit::Player(db.getDoubleFromRow(db.getStorageRefference(), "rating", row), db.getDoubleFromRow(db.getStorageRefference(), "stamina", row), db.getDoubleFromRow(db.getStorageRefference(), "age", row), db.getDoubleFromRow(db.getStorageRefference(), "experience", row), db.getStringFromRow(db.getStorageRefference(), "name", row), FieldUnit::Player::PositionType::DEFENDER));
				break;
			}
			case 1: {
				players.push_back(FieldUnit::Player(db.getDoubleFromRow(db.getStorageRefference(), "rating", row), db.getDoubleFromRow(db.getStorageRefference(), "stamina", row), db.getDoubleFromRow(db.getStorageRefference(), "age", row), db.getDoubleFromRow(db.getStorageRefference(), "experience", row), db.getStringFromRow(db.getStorageRefference(), "name", row), FieldUnit::Player::PositionType::STRIKERS));
				break;
			}
			case 2: {
				players.push_back(FieldUnit::Player(db.getDoubleFromRow(db.getStorageRefference(), "rating", row), db.getDoubleFromRow(db.getStorageRefference(), "stamina", row), db.getDoubleFromRow(db.getStorageRefference(), "age", row), db.getDoubleFromRow(db.getStorageRefference(), "experience", row), db.getStringFromRow(db.getStorageRefference(), "name", row), FieldUnit::Player::PositionType::MIDFIELDER));
				break;
			}
			default: abort();
			}
			row++;
		}
		layout /= 10;
	}
	return players;
}

std::vector<FieldUnit::Player> SoccerDecoder::FillPlayersMYSQL(sql::ResultSet* res, int layout)
{
	std::vector<FieldUnit::Player> players;
	res->first();
	for (int i = 0; i < 3; ++i) {
		int temp = layout % 10;
		for (int j = 0; j < temp; ++j) {
			switch (i) {
			case 0: {
				players.push_back(FieldUnit::Player(static_cast<double>(res->getDouble("rating")), static_cast<double>(res->getDouble("stamina")), static_cast<double>(res->getDouble("age")), static_cast<double>(res->getDouble("experience")), res->getString("name"), FieldUnit::Player::PositionType::DEFENDER));
				break;
			}
			case 1: {
				players.push_back(FieldUnit::Player(static_cast<double>(res->getDouble("rating")), static_cast<double>(res->getDouble("stamina")), static_cast<double>(res->getDouble("age")), static_cast<double>(res->getDouble("experience")), res->getString("name"), FieldUnit::Player::PositionType::STRIKERS));
				break;
			}
			case 2: {
				players.push_back(FieldUnit::Player(static_cast<double>(res->getDouble("rating")), static_cast<double>(res->getDouble("stamina")), static_cast<double>(res->getDouble("age")), static_cast<double>(res->getDouble("experience")), res->getString("name"), FieldUnit::Player::PositionType::MIDFIELDER));
				break;
			}
			default: abort();
			}
			res->next();
		}
		layout /= 10;
	}
	return players;
}

std::vector<int> SoccerDecoder::Chunks(const int count)
{
	int total = 0;
	std::vector<int> chunk;
	for (size_t i = 0; i < this->threadCount; i++) {
		chunk.push_back(std::floor(count / (this->threadCount + 1)));
		total += std::floor(count / (this->threadCount + 1));
	}
	if (std::floor(count) - total != 0)
		chunk.push_back(count - total);
	return chunk;
}

File: projects/PHP/classes/SoccerDecoder.php
<?php

include_once "Player.php";
include_once "Team.php";
include_once "TeamMatch.php";

set_error_handler(array('SoccerDecoder', 'errorHandlerAll'));

define('SQL_LAYOUTA', 'SELECT `layout` FROM `teams` WHERE `id` LIKE 1');
define('SQL_LAYOUTB', 'SELECT `layout` FROM `teams` WHERE `id` LIKE 2');
define('SQL_PLAYERSA', 'SELECT * FROM `players` WHERE `team_id` LIKE 1 ORDER BY `rating` ASC LIMIT 11');
define('SQL_PLAYERSB', 'SELECT * FROM `players` WHERE `team_id` LIKE 2 ORDER BY `rating` ASC LIMIT 11');
define('SQL_USE_DB', 'USE soccer');

class SoccerDecoder
{
    private object $conn;
    private int $faceoffs;
    private object $attackingTeam;
    private object $defendingTeam;
    private int $duration = 0;
    private int $aw = 0;
    private int $bw = 0;
    private int $dw = 0;

    public function __construct(int $faceoffs, int $mode, string $loc)
    {
        $layoutA = $layoutB = $playersA = $playersB = null;
        $this->faceoffs = $faceoffs;
        switch ($mode) {
            case 0:{
                    $this->conn = $this->SQLITE_Open($loc, SQLITE3_OPEN_READWRITE);
                    $layoutA = $this->SQLITE_Fetch($this->SQLITE_Execute($this->conn, SQL_LAYOUTA))['layout'];
                    $layoutB = $this->SQLITE_Fetch($this->SQLITE_Execute($this->conn, SQL_LAYOUTB))['layout'];
                    $playersA = $this->SQLITE_Fetch($this->SQLITE_Execute($this->conn, SQL_PLAYERSA));
                    $playersB = $this->SQLITE_Fetch($this->SQLITE_Execute($this->conn, SQL_PLAYERSB));
                    break;
                }
            case 1:
            case 2:{
                    $this->conn = new mysqli('localhost', 'root', 'securepassword', 'soccer');
                    $layoutA = $this->MYSQL_Fetch($this->conn, SQL_LAYOUTA)['layout'];
                    $layoutB = $this->MYSQL_Fetch($this->conn, SQL_LAYOUTB)['layout'];
                    $playersA = $this->MYSQL_Fetch($this->conn, SQL_PLAYERSA);
                    $playersB = $this->MYSQL_Fetch($this->conn, SQL_PLAYERSB);
                    break;
                }
            default: throw new Exception('SD :: Invalid launch option');
        }

        assert(count($playersA) - 1 == 10 && count($playersB) - 1 == 10);
        
        $this->setAttackingTeam(new \Contest\Team("Attackers", \Contest\FieldType::AWAY, $this->fillPlayers($playersA, $layoutA), new \FieldUnit\Player($playersA[10]['rating'], $playersA[10]['stamina'], $playersA[10]['age'], $playersA[10]['experience'], $playersA[10]['name'], $playersA[10]['position']), $layoutA));
        $this->setDefendingTeam(new \Contest\Team("Defenders", \Contest\FieldType::HOME, $this->fillPlayers($playersB, $layoutB), new \FieldUnit\Player($playersB[10]['rating'], $playersB[10]['stamina'], $playersB[10]['age'], $playersB[10]['experience'], $playersB[10]['name'], $playersB[10]['position']), $layoutB));
    }

    private function fillPlayers($result, $layout) {
        $row = 0;
        $players = [];
        for ($i = 0; $i < 3; ++$i) {
            $temp = $layout % 10;
            for ($j = 0; $j < $temp; ++$j) {
                switch ($i) {
                    case 0:
                        {
                            array_push($players, new \FieldUnit\Player($result[$row]['rating'], $result[$row]['stamina'], $result[$row]['age'], $result[$row]['experience'], $result[$row]['name'], \FieldUnit\PositionType::DEFENDER));
                            break;
                        }
                    case 1:
                        {
				            array_push($players, new \FieldUnit\Player($result[$row]['rating'], $result[$row]['stamina'], $result[$row]['age'], $result[$row]['experience'], $result[$row]['name'], \FieldUnit\PositionType::STRIKER));
                            break;
                        }
                    case 2:
                        {
                            array_push($players, new \FieldUnit\Player($result[$row]['rating'], $result[$row]['stamina'], $result[$row]['age'], $result[$row]['experience'], $result[$row]['name'], \FieldUnit\PositionType::MIDFIELDER));
                            break;
                        }
                    default:
                        throw new Exception("Fatal error");
                }
                $row++;
            }
            $layout /= 10;
        }
        return $players;
    }

    public function start(){
        srand((function(){ 
            list($usec, $sec) = explode(' ', microtime());
            return $sec + $usec * 1000000;
        })());
        $startTime = microtime(true);
        for($i = 0; $i < $this->faceoffs; ++$i){
            $faceoff = new \Contest\FaceOff(clone $this->attackingTeam, clone $this->defendingTeam);
            $faceoff->setupMatch();
            $faceoff->simulate();
            if ($faceoff->getAttackingTeam()->getScore() > $faceoff->getDefendingTeam()->getScore()) $this->aw++;
            else if ($faceoff->getAttackingTeam()->getScore() > $faceoff->getDefendingTeam()->getScore()) $this->bw++;
            else $this->dw++;
        }
        $this->setDuration(( microtime(true) - $startTime ) * 1000);
    }

    public function printResultJson(){
        echo "{\"a\":" . $this->aw . "," . "\"b\":" . $this->bw . "," . "\"draws\":" . $this->dw . "," . "\"duration\":" . $this->getDuration() . "}";
    }

    public static function errorHandlerAll(int $errno, string $message, string $file, int $line)
    {
        if($errno === E_WARNING) {
            $errno = E_ERROR;
        }
        return false;
    }

    private function SQLITE_Open(string $location, int $mode)
    {
        $handle = new SQLite3($location, $mode);
        if (!$handle) {
            throw new Exception('SQLITE :: Connection error: ' . $handle->lastErrorMsg());
        }
        return $handle;
    }

    private function SQLITE_Execute(object $conn, string $query)
    {
        $array['dbhandle'] = $conn;
        $array['query'] = $query;
        $result = $conn->query($query);
        return $result;
    }

    private function SQLITE_Fetch(object $result)
    {
        $out = array();
        while ($row = $result->fetchArray(SQLITE3_ASSOC)) {
            $out[] = $row;
        }
        if(count($out) == 1) return $out[0];
        return $out;
    }

    private function MYSQL_Open(string $endpoint, string $username, string $password, string $db)
    {
        $handle = new mysqli($endpoint, $username, $password, $db);
        if ($handle->connect_error) {
            throw new Exception('MYSQL :: Connection error: ' . $handle->connect_error);
        }
        return $handle;
    }

    private function MYSQL_Execute(object $conn, string $sql, mixed $types = null, mixed ...$params)
    {
        if (!($stmt = $conn->prepare($sql))) {
            throw new Exception('MYSQL :: Failed to prepare statement');
            exit(-1);
        }

        if($types != null){
            if (strlen($types) > 0) {
                $bindParamArray = array();
                $bindParamArray[] = &$types;
                for ($i = 0; $i < strlen($types); ++$i) {
                    $bindParamArray[] = &$params[$i];
                }
    
                if (!(call_user_func_array(array($stmt, 'bind_param'), $bindParamArray))) {
                    $stmt->close();
                    throw new Exception('MYSQL :: Failed to call_user_func_array on bind_param');
                    exit(-1);
                }
            }
        }

        if (!($stmt->execute())) {
            $stmt->close();
            throw new Exception('MYSQL :: Failed to execute query');
            exit(-1);
        }

        $stmt->close();
        return true;
    }

    private function MYSQL_Fetch(object $conn, string $sql, mixed $types = null, mixed ...$params)
    {
        $out = array();

        if (!($stmt = $conn->prepare($sql))) {
            throw new Exception('MYSQL :: Failed to prepare statement');
            exit(-1);
        }

        if($types != null){
            if (strlen($types) > 0) {
                $bindParamArray = array();
                $bindParamArray[] = &$types;
                for ($i = 0; $i < strlen($types); ++$i) {
                    $bindParamArray[] = &$params[$i];
                }
    
                if (!(call_user_func_array(array($stmt, 'bind_param'), $bindParamArray))) {
                    $stmt->close();
                    throw new Exception('MYSQL :: Failed to call_user_func_array on bind_param');
                    exit(-1);
                }
            }
        }

        if (!($stmt->execute())) {
            $stmt->close();
            throw new Exception('MYSQL :: Failed to execute query');
            exit(-1);
        }

        if (!($result = $stmt->get_result())) {
            $stmt->close();
            throw new Exception('MYSQL :: Failed to get query result');
            exit(-1);
        }
        while ($row = $result->fetch_array(MYSQLI_ASSOC)) {
            $out[] = $row;
        }

        $stmt->close();

        if(count($out) == 1) return $out[0];
        return $out;
    }

    /**
     * Get the value of duration
     */ 
    public function getDuration()
    {
        return $this->duration;
    }

    /**
     * Set the value of duration
     *
     * @return  self
     */ 
    public function setDuration($duration)
    {
        $this->duration = $duration;
    }

    /**
     * Get the value of attackingTeam
     */ 
    public function getAttackingTeam()
    {
        return $this->attackingTeam;
    }

    /**
     * Set the value of attackingTeam
     *
     * @return  self
     */ 
    public function setAttackingTeam($attackingTeam)
    {
        $this->attackingTeam = $attackingTeam;
    }

    /**
     * Get the value of defendingTeam
     */ 
    public function getDefendingTeam()
    {
        return $this->defendingTeam;
    }

    /**
     * Set the value of defendingTeam
     *
     * @return  self
     */ 
    public function setDefendingTeam($defendingTeam)
    {
        $this->defendingTeam = $defendingTeam;
    }
}
?>